integration/util.py

```diff
@@ -593,87 +596,74 @@ def _check_status(response):
     _check_status(resp)
-    return resp.content
+    body = yield resp.content()
+    returnValue(body)
 
 
-def web_post(tahoe, uri_fragment, **kwargs):
-    """
-    Make a POST request to the webport of `node` (a `TahoeProcess,
-    usually from a fixture e.g. `alice`). This will look like:
-    `http://localhost:<webport>/<uri_fragment>`. All `kwargs` are
-    passed on to `requests.post`
-    """
-    url = node_url(tahoe.node_dir, uri_fragment)
-    resp = requests.post(url, **kwargs)
-    _check_status(resp)
-    return resp.content
+def twisted_sleep(reactor, timeout):
+    return deferLater(reactor, timeout, lambda: None)
 
 
-def await_client_ready(tahoe, timeout=10, liveness=60*2):
+@inlineCallbacks
+def await_client_ready(reactor, tahoe, timeout=10, liveness=60*2):
     """
     Uses the status API to wait for a client-type node (in `tahoe`, a
     `TahoeProcess` instance usually from a fixture e.g. `alice`) to be
     'ready'. A client is deemed ready if:
 
       - it answers `http://<node_url>/statistics/?t=json/`
       - there is at least one storage-server connected
-      - every storage-server has a "last_received_data" and it is
-        within the last `liveness` seconds
+      - it has a "last_received_data" within the last `liveness` seconds
 
     We will try for up to `timeout` seconds for the above conditions
     to be true. Otherwise, an exception is raised
     """
-    start = time.time()
-    while (time.time() - start) < float(timeout):
+    start = reactor.seconds()
+    while (reactor.seconds() - start) < float(timeout):
         try:
-            data = web_get(tahoe, u"", params={u"t": u"json"})
+            data = yield web_get(tahoe, u"", params={u"t": u"json"})
             js = json.loads(data)
         except Exception as e:
             print("waiting because '{}'".format(e))
-            time.sleep(1)
+            yield twisted_sleep(reactor, 1)
             continue
 
         if len(js['servers']) == 0:
             print("waiting because no servers at all")
-            time.sleep(1)
+            yield twisted_sleep(reactor, 1)
             continue
         server_times = [
             server['last_received_data']
             for server in js['servers']
+            if server['last_received_data'] is not None
         ]
-        # if any times are null/None that server has never been
-        # contacted (so it's down still, probably)
-        if any(t is None for t in server_times):
-            print("waiting because at least one server not contacted")
-            time.sleep(1)
-            continue
 
         # check that all times are 'recent enough'
-        if any([time.time() - t > liveness for t in server_times]):
-            print("waiting because at least one server too old")
-            time.sleep(1)
+        if all([time.time() - t > liveness for t in server_times]):
+            print("waiting because no server new enough")
+            yield twisted_sleep(reactor, 1)
```

Comment:
Because `happy=1`. So, since we're using a single Introducer .. any client that is started (with storage) but then stopped will no longer report a connection here at all (i.e. their time will be `None`). Also the behavior described in the docstring wasn't what this was actually doing.

So basically a client is "ready" if there are enough recent storage-servers so that a Tahoe operation will succeed. In this case, 1.